internal/exec/stages/files/units.go

```diff
@@ -17,37 +17,96 @@ package files
 import (
 	"fmt"
 	"path/filepath"
+	"strconv"
 	"strings"
 
+	"github.com/coreos/go-systemd/dbus"
 	"github.com/coreos/ignition/v2/config/v3_1_experimental/types"
 	"github.com/coreos/ignition/v2/internal/distro"
 	"github.com/coreos/ignition/v2/internal/exec/util"
 )
 
+type Preset struct {
+	unit            string
+	enabled         bool
+	instantiateable bool
+	instances       []string
+}
+
+// warnOnOldSystemdVersion checks the version of Systemd
+// in a given system.
+func (s *stage) warnOnOldSystemdVersion() error {
+	var systemdVersion string
+	var begin, end int
+	conn, err := dbus.New()
+	if err != nil {
+		return err
+	}
+	version, err := conn.GetManagerProperty("Version")
+	if err != nil {
+		return err
+	}
+	// Handle different systemd versioning schemes that are being returned.
+	// for e.g:
+	// - Fedora 31: `v243.5-1.fc31`
+	// - RHEL 8: `239`
+	if strings.HasPrefix(version, "v") {
+		begin = 1
+
+	} else {
+		begin = 0
+	}
+	end = strings.Index(version, ".")
+	if end == -1 {
+		end = len(version)
+	}
+	systemdVersion = version[begin:end]
+	if value, err := strconv.Atoi(systemdVersion); err == nil {
+		if value < 240 {
+			if err := s.Logger.Warning("The version of systemd (%q) is less than 240. Enabling/disabling instantiated units may not work. See https://github.com/coreos/ignition/issues/586 for more information.", systemdVersion); err != nil {
+				return err
+			}
+		}
+	} else {
+		return err
+	}
+	return nil
+}
+
 // createUnits creates the units listed under systemd.units.
 func (s *stage) createUnits(config types.Config) error {
-	enabledOneUnit := false
+	presets := make(map[string]*Preset)
 	for _, unit := range config.Systemd.Units {
 		if err := s.writeSystemdUnit(unit, false); err != nil {
 			return err
 		}
 		if unit.Enabled != nil {
 			if *unit.Enabled {
-				if err := s.Logger.LogOp(
-					func() error { return s.EnableUnit(unit) },
-					"enabling unit %q", unit.Name,
-				); err != nil {
-					return err
+				if strings.Contains(unit.Name, "@") {
+					unitName, instance := parseInstanceUnit(unit)
+					if len(instance) > 0 {
+						if _, ok := presets[unitName]; ok {
+							presets[unitName].instances = append(presets[unitName].instances, instance)
+						} else {
+							presets[unitName] = &Preset{unitName, true, true, []string{instance}}
+						}
+					}
```

Comment:
we shouldn't need this to be inside an if condition `instance` should always be a string > 0 length, right?

Reply:
same for the logic below that follows this same path